pyuvdata/tests/__init__.py

```diff
@@ -5,19 +5,245 @@
+
+        # Built-in catch_warnings does not reset entered state so we do it
+        # manually here for this context manager to become reusable.
+        self._entered = False
+
+        # only check if we're not currently handling an exception
+        if exc_type is None and exc_val is None and exc_tb is None:
+            if self.expected_warning is not None:
+                assert len(self) == len(self.expected_warning), (
+                    f"{len(self.expected_warning)} warnings expected, "
+                    f"{len(self)} warnings issued. The list of emitted warnings is: "
+                    f"{[each.message for each in self]}."
+                )
+
+                for warn_i, exp_warn in enumerate(self.expected_warning):
+                    if not any(issubclass(r.category, exp_warn) for r in self):
+                        __tracebackhide__ = True
+                        raise AssertionError(
+                            "DID NOT WARN. No warnings of type {} was emitted. "
+                            "The list of emitted warnings is: {}.".format(
+                                self.expected_warning, [each.message for each in self]
+                            )
+                        )
+                    elif self.match is not None:
+                        for record in self:
+                            if str(record.message).startswith(
+                                self.match[warn_i]
+                            ) or re.compile(self.match[warn_i]).search(
+                                str(record.message)
+                            ):
+                                if issubclass(
+                                    record.category, self.expected_warning[warn_i]
+                                ):
+                                    break
+                        else:
+                            raise AssertionError(
+                                f"No warnings of type {self.expected_warning[warn_i]} "
+                                f"matching ('{self.match[warn_i]}') was "
+                                "emitted. The list of emitted warnings is: "
+                                f"{[each.message for each in self]}."
+                            )
+
+
+def check_warnings(expected_warning, match, nwarnings=None, *args, **kwargs):
+    """
+    Assert that code raises a particular set of warnings, used as a context manager.
+
+    Similar to ``pytest.warns``, but allows for specifying multiple warnings.
+    It also better matches warning strings when the warning uses f-strings or
+    formating.
+
+    Note that unlike the older checkWarnings function, the warnings can be passed
+    in any order, they do not have to match the order the warnings are raised
+    in the code.
+
+    Parameters
+    ----------
+    expected_warning : list of Warnings or Warning
+        List of expected warnings. If a single warning type or a length 1 list,
+        will be used for the type of all warnings.
```

Comment:
I may have to go back and check, but it seems like you only need a list of all the unique types of warnings created by a test since for each record, the list of expected types is looped through until a matching one is found.

Reply:
I thought I was writing it so that the message and error type had to match (so that I use the message index to get the error index I'm checking against), but I may have made a mistake.

Reply:
Further inspection looks like I missed a `warn_i` index into the `expected_warnings` list. should work the way you intend I think.